Marlin/src/HAL/HAL_DUE/HAL_Due.h

```diff
@@ -0,0 +1,139 @@
+ (at your option) any later version.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+****************************************************************************/
+
+/**
+ * Description: HAL for Arduino Due and compatible (SAM3X8E)
+ *
+ * For ARDUINO_ARCH_SAM
+ */
+
+
+#ifndef _HAL_DUE_H
+#define _HAL_DUE_H
+
+// --------------------------------------------------------------------------
+// Includes
+// --------------------------------------------------------------------------
+
+#include <stdint.h>
+
+#include "Arduino.h"
+
+#include "fastio_Due.h"
+#include "watchdog_Due.h"
+
+
+
+// --------------------------------------------------------------------------
+// Defines
+// --------------------------------------------------------------------------
+
+#define _BV(bit) 	(1 << (bit))
+
+#define analogInputToDigitalPin(IO) IO
+
+#define     CRITICAL_SECTION_START	uint32_t primask=__get_PRIMASK(); __disable_irq();
+#define     CRITICAL_SECTION_END    if (primask==0) __enable_irq();
+
+// On AVR this is in math.h?
+#define square(x) ((x)*(x))
+
+#define strncpy_P(dest, src, num) strncpy((dest), (src), (num))
+
+// --------------------------------------------------------------------------
+// Types
+// --------------------------------------------------------------------------
+
+#define HAL_TIMER_TYPE unsigned long
+
+// --------------------------------------------------------------------------
+// Public Variables
+// --------------------------------------------------------------------------
+
+// reset reason set by bootloader
+extern uint8_t MCUSR;
+
+// --------------------------------------------------------------------------
+// Public functions
+// --------------------------------------------------------------------------
+
+// Disable interrupts
+void cli(void);
+
+// Enable interrupts
+void sei(void);
+
+void _delay_ms(int delay);
+
+int freeMemory(void);
+
+// SPI: Extended functions which take a channel number (hardware SPI only)
+/** Write single byte to specified SPI channel */
+void spiSend(uint32_t chan, byte b);
+/** Write buffer to specified SPI channel */
+void spiSend(uint32_t chan, const uint8_t* buf, size_t n);
+/** Read single byte from specified SPI channel */
+uint8_t spiRec(uint32_t chan);
+
+
+
+// EEPROM
+void eeprom_write_byte(unsigned char *pos, unsigned char value);
+unsigned char eeprom_read_byte(unsigned char *pos);
+void eeprom_read_block (void *__dst, const void *__src, size_t __n);
+void eeprom_update_block (const void *__src, void *__dst, size_t __n);
+
+
+// timers
+#define STEP_TIMER_NUM 3
+#define TEMP_TIMER_NUM 4
+
+#define HAL_TIMER_RATE 		     (F_CPU/32.0)
```

Comment:
This will work on both systems. The problem will occure here: https://github.com/MarlinFirmware/Marlin/blob/71ad241b6922f33fff753a33a218e2ec28302fe2/Marlin/stepper.h#L320

That was 2 years ago the difference between 40k steps/s and 300k steps/s.